protocol/codec_tester.go

```diff
@@ -433,7 +433,7 @@ func RunEncodingTest(t *testing.T, template msgpMarshalUnmarshal) {
 
 		// some objects might appen to the original error additional info.
 		// we ensure that invalidObject error is not failing the test.
-		if errors.As(err, &ErrInvalidObject) {
+		if errors.Is(err, ErrInvalidObject) {
```

Comment:
@algoidan I think you wrote this originally? The linter went off saying:
```
protocol/codec_tester.go:436:6: errorsas: second argument to errors.As should not be *error (govet)
		if errors.As(err, &ErrInvalidObject) {
		   ^
```
So I think I fixed it here, but I'm curious because I think that means the error check didn't work before?

Reply:
I think it would, wouldn't it? It would just _also_ assign the error object into the global var `ErrInvalidObject` which probably wasn't intended but seems like it wouldn't break things.

Reply:
Yeah ErrInvalidObject is not a type, and errors.As is supposed to be used for error types, not error values ... and I think this means that before, it was assigning `err` the pointer &ErrInvalidObject which led to side effects (???)

> errors.As finds the first error in err's chain that matches target, and if one is found, sets target to that error value and returns true. Otherwise, it returns false.
> An error matches target if the error's concrete value is assignable to the value pointed to by target, or if the error has a method As(interface{}) bool such that As(target) returns true. In the latter case, the As method is responsible for setting target.

And now a ton of tests are failing CI because of this line ...

Reply:
I just gave a try fixing this closer to the source, by having the randomizeValue function only generate valid HashType (I believe it is the only custom msgp deserializer implementation that raises errors on random values, and ErrInvalidObject was created solely for crypto.HashType used by state proofs). It works but may seem a little hacky, probably the more elegant solution is to pass the MaxHashType via tag into codec_tester

Reply:
Found this comment: https://github.com/algorand/go-algorand/pull/3097#discussion_r733760179

Chris' change looks correct to me. The old code should technically work, but it's weird since multiple references to the same error would reassign themselves to themself.

Reply:
Oh yes, I see, the same concern as govet had — I think I like my fix better than skipping these errors because you don't get to actually test random encodings if the deserializer is erroring out once you hit this field the vast majority of the time (it will only be valid for 3 out of 2^16 runs, right, so the randomized encoding tests are going to be barely ever exercised the way it is working now?).
![image](https://user-images.githubusercontent.com/51567/184935932-17cbd03c-f93b-467e-972f-8839f0209744.png)

Reply:
@winder That's what I thought at first too, that it was ok, but would weirdly assign to itself over and over. But I think it's much worse.  Since the type of `&ErrInvalidObject` is `*error`, rather than something specific like `*specialErrType`, the test that `As` performs should _always_ succeed.  So if you ever tried:
`errors.As(fmt.Errorf("junk"), &ErrInvalidObject)` I'm pretty sure it would be true.

Reply:
Yes, it looks like this bug was causing any msgp decode error to be ignored before like JJ mentioned: https://go.dev/play/p/CYe-zWl2zHV

My fix (to generate valid HashType uint16 fields) is showing me new issues that were hiding behind this, related to pointer types not being supported by the codec_tester:
```
Failed
=== RUN   TestRandomizedEncodingStateProofKeys
    codec_tester.go:438: 
        	Error Trace:	codec_tester.go:438
        	            				msgp_gen_test.go:99
        	Error:      	Received unexpected error:
        	            	unsupported object kind ptr
        	Test:       	TestRandomizedEncodingStateProofKeys
--- FAIL: TestRandomizedEncodingStateProofKeys (0.00s)
```
due to
```go
    KeyRoundPair struct {
        _struct struct{} `codec:",omitempty,omitemptyarray"`

        Round uint64               `codec:"rnd"`
        Key   *crypto.FalconSigner `codec:"key"`
    }
```

Reply:
OK, I fixed this by updating the codec_tester to be able to generate pointers to random structs.